### PR TITLE
(1901) Integrate with Okta for user login

### PR DESCRIPTION
**Pull request by pezholio**
This uses the Okta Open ID Middleware (https://github.com/okta/okta-oidc-middleware) to integrate our app with Okta and
add simple login / logout functionality. It's not (yet) linked to a real user in our application, but this can come later.

### CODE REVIEW

File: src/authentication/authentication.guard.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * it is true or false.
+ *
+ * @example (in a controller, as metadata above a controller action)
+ *
+ * @UseGuards(AuthenticationGuard)
+ *
+ */
+@Injectable()
+export class AuthenticationGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
```
Comment (bryangaledxw) [inline src/authentication/authentication.guard.ts:22]: Do we need reflector here?

Comment (pezholio) [reply src/authentication/authentication.guard.ts:22]: Ah, no we don't. I think I cargo culted that from somewhere else. Will fix 😃

File: src/app.controller.ts [hunk 2]
```diff
@@ -10,4 +20,18 @@ export class AppController {
...
+  @Render('admin')
+  admin(): object {
+    return {};
+  }
+
+  @Get('/logout')
+  @UseGuards(AuthenticationGuard)
+  @Redirect('/')
+  logout(@Req() req: Request): object {
+    return req.logout();
```
Comment (bryangaledxw) [inline src/app.controller.ts:35]: This doesn't seem to perform a logout in the "user has to re-enter their password with Okta" sense. Is there some equivalent to this we want - https://github.com/panva/node-openid-client/blob/main/docs/README.md#clientendsessionurlparameters - a URL to redirect the user to that will perform an Okta-side logout?

Comment (bryangaledxw) [reply src/app.controller.ts:35]: Hmm, the more I play with this the more it doesn't seem like a reliable logout? If I visit logout, then admin, then admin a second time, I'm logged in

Comment (bryangaledxw) [reply src/app.controller.ts:35]: Ah n/m, that's the redirect to login doing its thing

File: src/main.ts [hunk 3]
```diff
@@ -50,6 +61,8 @@ async function bootstrap() {
     }),
   );
 
+  app.useGlobalFilters(new UnauthorizedExceptionFilter());
+
   await app.listen(3000);
```
Comment (bryangaledxw) [inline src/main.ts:66]: Going by the docs, we want something like:
`await new Promise((res, rej) => {oidc.on('ready', res); oidc.on('error', rej);}); `

before starting the server

File: src/app.controller.ts (hunk 2 above)
Comment (bryangaledxw) [reply src/app.controller.ts:35]: Ah k, see - https://github.com/okta/okta-oidc-middleware#oidcforcelogoutandrevoke

I think this is what we want instead? Though note the caution on logout on GET. Though also, the middleware provides a built in POST 'logout' route

Comment (pezholio) [reply src/app.controller.ts:35]: Ha, nice. I think you're right though - we should really be `POST`ing for a logout

File: src/main.ts (hunk 3 above)
Comment (pezholio) [reply src/main.ts:66]: Good catch. I think this is what might have been causing those errors we were seeing too.

File: cypress/integration/application.spec.js [hunk 4]
```diff
@@ -3,4 +3,22 @@ describe('/', () => {
...
+
+      cy.get('body').should('contain', 'Hello Admin!');
+    });
+
+    it('allows me to logout', () => {
+      cy.visit('/admin');
+
+      cy.get('[data-purpose=logout]').click();
+    });
+  });
```
Comment (bryangaledxw) [inline cypress/integration/application.spec.js:23]: Should we test that admin redirects when if accessed when not logged in, and that logout does actually log the user out?

File: cypress/support/commands.js [hunk 5]
```diff
@@ -1,25 +1,49 @@
...
-// Cypress.Commands.add('drag', { prevSubject: 'element'}, (subject, options) => { ... })
-//
-//
-// -- This is a dual command --
-// Cypress.Commands.add('dismiss', { prevSubject: 'optional'}, (subject, options) => { ... })
-//
-//
-// -- This will overwrite an existing command --
-// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+Cypress.Commands.add('login', () => {
```
Comment (bryangaledxw) [inline cypress/support/commands.js:1]: So I understand why we need this, is this so we're not depending relying on the HTML of the Okta login page not changing? Am I right in thinking Cypress _could_ just populate the Okta login page with a username and password?

File: src/app.controller.ts [hunk 6]
```diff
@@ -10,4 +12,11 @@ export class AppController {
   getHello(): object {
     return {};
   }
+
+  @Get('/admin')
+  @UseGuards(AuthenticationGuard)
```
Comment (bryangaledxw) [inline src/app.controller.ts:17]: Again for my understanding, I guess down the line we want a guard that checks not just that the user is logged in, but that they actually have admin permissions. Or do you see that being logic in the function body?

File: src/main.ts [hunk 7]
```diff
@@ -42,6 +44,15 @@ async function bootstrap() {
   app.setBaseViewsDir(views);
   app.setViewEngine('njk');
 
+  app.use(
+    require('express-session')({
```
Comment (bryangaledxw) [inline src/main.ts:48]: Can we install @types/express-session and do `import * as session from 'express-session';` to keep things Typescript-y?

File: views/admin.njk [hunk 8]
```diff
@@ -0,0 +1,26 @@
+{% extends "base.njk" %}
+
+{% block header %}
+  {{
+    govukHeader({
+      navigation: [
+          {
+            html: '<form method="post" action="/logout">
```
Comment (bryangaledxw) [inline views/admin.njk:8]: Is it worth adding logout to the standard header, not just for the admin page (only shown if logged in) as part of this PR?

File: src/main.ts (hunk 7 above)
Comment (bryangaledxw) [reply src/main.ts:48]: Also, should we have a ticket somewhere to choose a production express-session store? See note about MemoryStore - https://github.com/expressjs/session

File: cypress/support/commands.js (hunk 5 above)
Comment (pezholio) [reply cypress/support/commands.js:1]: Yeah, it's considered best practice in Cypress land to not test anything outside of our domain. If we wanted to be _really_ clever, we could set up our own OAuth server, or stub out all the calls that we expect our server to make, but this seems likie a good half way house, especially now we can store and reuse the session at will throughout the suite.

File: src/app.controller.ts (hunk 6 above)
Comment (pezholio) [reply src/app.controller.ts:17]: Yeah, I think we'll have that as a guard too. We can set metadata to tell the guard what permissions a user needs too. There's some docs on this use case here https://docs.nestjs.com/guards#role-based-authentication

File: src/main.ts (hunk 7 above)
Comment (pezholio) [reply src/main.ts:48]: Both excellent shouts - I'm new to this world, so anything like this that you spot, let me know 👍

File: views/admin.njk (hunk 8 above)
Comment (pezholio) [reply views/admin.njk:8]: I'm leaning towards no for this. I'm seeing the two sections of the site as distinct and seperate, so having seperate layouts gives us flexibility. I could be wrong though!

**Comment by pezholio** [comment]
Closing this in favour of building something a tad more generic.

**Pull request closed without merging**